plugins/DRTubesSDAction.cpp

```diff
@@ -0,0 +1,267 @@
+  auto cpNo = aStep->GetPreStepPoint()->GetTouchable()->GetCopyNumber();
+  // The second bit of the CopyNumber corresponds to the "core" entry:
+  // 1 if the step is in the fiber core (S or C) and 0 if it is
+  // in the fiber cladding (C only) 
+  unsigned int CoreID = (cpNo & 0b10) >> 1; // take CpNo 2nd bit
+  bool IsCherClad = (CoreID == 0);
+  // The first bit of the CopyNumber corresponds to the "cherenkov" entry
+  // 1 for C fibers and 0 for S fibers
+  unsigned int CherenkovID = cpNo & 0b1;
+  bool IsScin = (CherenkovID == 0);
+  // Skip this step if edep is 0 and it is a scintillating fiber
+  if (IsScin && Edep == 0.) return true;
+  // If it is a track inside the cherenkov CLADDING skip this step,
+  // if it is an optical photon kill it first
+  if (IsCherClad) {
+    if (aStep->GetTrack()->GetParticleDefinition() == G4OpticalPhoton::Definition()) {
+      aStep->GetTrack()->SetTrackStatus(fStopAndKill);
+    }
+    return true;
+  }
+
+  // Now we are inside fibers' core volume (either Scintillating or Cherenkov)
+
+  // We recreate the TubeID from the tube copynumber:
+  // fist16 bits for the columnID and second 16 bits for the rowID
+  auto TubeID = aStep->GetPreStepPoint()->GetTouchable()->GetCopyNumber(2);
+  unsigned int ColumnID = TubeID >> 16;
+  unsigned int RawID = TubeID & 0xFFFF;
+  auto TowerID = static_cast<unsigned int>(aStep->GetPreStepPoint()->GetTouchable()->GetCopyNumber(3));
+  auto StaveID = static_cast<unsigned int>(aStep->GetPreStepPoint()->GetTouchable()->GetCopyNumber(4));
+
+  VolumeID VolID = 0; // recreate the 64-bit VolumeID
+  BitFieldCoder bc("stave:10,tower:6,air:1,col:16,row:16,clad:1,core:1,cherenkov:1");
+  bc.set(VolID, "stave" , StaveID);
+  bc.set(VolID, "tower" , TowerID);
+  bc.set(VolID, "air", 0);
+  bc.set(VolID, "col", ColumnID);
+  bc.set(VolID, "row", RawID);
+  bc.set(VolID, "clad", 1);
+  bc.set(VolID, "core", CoreID);
+  bc.set(VolID, "cherenkov", CherenkovID);
+
+  bool IsRight = (aStep->GetPreStepPoint()->GetPosition().z() > 0.);
+
+  // We now calculate the signal in S and C fiber according to the step contribution
+  //
+  G4double steplength = aStep->GetStepLength();
+  G4int signalhit = 0;
+  Geant4HitCollection* coll = (IsRight && IsScin)    ? collection(m_collectionID)
+                              : (IsRight && !IsScin) ? collection(m_userData.collection_cher_right)
+                              : (!IsRight && IsScin) ? collection(m_userData.collection_scin_left)
+                                                     : collection(m_userData.collection_cher_left);
+
+  if (IsScin) {  // it is a scintillating fiber
+
+    //m_userData.fEvtAction->AddEdepScin(Edep);
+
+    if (aStep->GetTrack()->GetDefinition()->GetPDGCharge() == 0 || steplength == 0.) {
+      return true;  // not ionizing particle
+    }
+    G4double distance_to_sipm = DRTubesSglHpr::GetDistanceToSiPM(aStep);
+    signalhit =
+      DRTubesSglHpr::SmearSSignal(DRTubesSglHpr::ApplyBirks(Edep, steplength));
```

Comment:
Are you sure about the application to the scintillation process only, and not to the energy deposit? I am pretty sure that this affected the deposited energy in the past, and this is what CALICE scintillator test beam simulations used, and what went into the scintillator calo simulation in DD4hep as well...

https://github.com/AIDASoft/DD4hep/blob/c9023f116f4f161f21d244fccdbc971753bd748f/DDG4/src/Geant4StepHandler.cpp#L90

Reply:
I am reviewing the Geant4 code for Birks Law right now and there is no trace of something that could affect the energy deposited. I also discussed with Geant4 colleagues, Briks Law in Geant4 can only affect the light emission process but not the energy deposited in the step. Of course surprises can happen, if you have evidence of something different I would be interested.

Reply:
This code in geant4 https://github.com/Geant4/geant4/blob/e58e650b32b961c8093f3dd6a2c3bc917b2552be/source/processes/electromagnetic/utils/src/G4EmSaturation.cc#L79 that is called from the step handler in DD4hep.

Reply:
Ok I'm learning something here, unfortunately in Geant4 we have a lot of code that should not be exposed to users. If I understand it correctly, the Geant4StepHandler applies Birks Law to every step (or at least every time it is applied to a G4Step in the SDAction code). The point is, for almost every material (step) it will find that Birks Constant is 0 (default) and do absolutely nothing. It seems a very inefficient thing to do. Why not let the users define the Birks function in their sensitive detector code for their specific materials? This is the standard I encountered so far in LHC experiments simulations. Anyhow, in my case this is not a problem because in the geometry materials Briks Constants are never set and in my SDAction I do not use Geant4StepHandler... We could also have a chat in person about this.

Reply:
No, the step handler is only called by the Scintillator Calo SensitiveDetector.
https://github.com/AIDASoft/DD4hep/blob/c9023f116f4f161f21d244fccdbc971753bd748f/DDG4/plugins/Geant4SDActions.cpp#L445 So it will be applied selectively to steps in the sensitive material, and Birks constant should be defined for the sensitive material, otherwise a different sensitive detector could be chosen.

Why not re-use what is available in Geant4 for Applying Birks' Law though? Which is a more complete implementation than what you have?

But it is true that we don't apply Birks' Law twice accidentally because it comes via the sensitive detector code.